# Deep Neural Network

# Loan Prediction

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sn

### Load and preprocessing data

In [2]:
data = pd.read_csv('train_loan_prediction.csv')
data.head()

Loan_ID Gender Married Dependents     Education Self_Employed  \
0  LP001002   Male      No          0      Graduate            No   
1  LP001003   Male     Yes          1      Graduate            No   
2  LP001005   Male     Yes          0      Graduate           Yes   
3  LP001006   Male     Yes          0  Not Graduate            No   
4  LP001008   Male      No          0      Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0         NaN             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History Property_Area Loan_Status  
0             1.0         Urban           Y  
1             1.0         Rural           N  
2             1.0         Urban           Y  
3             1.0         Urban           Y  
4             1.0         Urban           Y

In [3]:
data.shape

(614, 13)

In [4]:
data.isnull().sum().sum()

149

In [5]:
print(f" There is {data.isnull().sum().sum()} missing values in this dataset")

 There is 149 missing values in this dataset


In [6]:
#let's see the type of data each variable has.
data.info

<bound method DataFrame.info of       Loan_ID  Gender Married Dependents     Education Self_Employed  \
0    LP001002    Male      No          0      Graduate            No   
1    LP001003    Male     Yes          1      Graduate            No   
2    LP001005    Male     Yes          0      Graduate           Yes   
3    LP001006    Male     Yes          0  Not Graduate            No   
4    LP001008    Male      No          0      Graduate            No   
..        ...     ...     ...        ...           ...           ...   
609  LP002978  Female      No          0      Graduate            No   
610  LP002979    Male     Yes         3+      Graduate            No   
611  LP002983    Male     Yes          1      Graduate            No   
612  LP002984    Male     Yes          2      Graduate            No   
613  LP002990  Female      No          0      Graduate           Yes   

     ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0               5849          

In [7]:
data.isnull().sum()

Loan_ID               0
Gender               13
Married               3
Dependents           15
Education             0
Self_Employed        32
ApplicantIncome       0
CoapplicantIncome     0
LoanAmount           22
Loan_Amount_Term     14
Credit_History       50
Property_Area         0
Loan_Status           0
dtype: int64

In [8]:
data = data.drop('Loan_ID', axis =1)
data.head()

Gender Married Dependents     Education Self_Employed  ApplicantIncome  \
0   Male      No          0      Graduate            No             5849   
1   Male     Yes          1      Graduate            No             4583   
2   Male     Yes          0      Graduate           Yes             3000   
3   Male     Yes          0  Not Graduate            No             2583   
4   Male      No          0      Graduate            No             6000   

   CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
0                0.0         NaN             360.0             1.0   
1             1508.0       128.0             360.0             1.0   
2                0.0        66.0             360.0             1.0   
3             2358.0       120.0             360.0             1.0   
4                0.0       141.0             360.0             1.0   

  Property_Area Loan_Status  
0         Urban           Y  
1         Rural           N  
2         Urban           Y  
3         Urban           Y  
4         Urban           Y

## Cleaning data

we will replace some of the missing values with the sub-variables that have the most count.

### Gender

In [9]:
data['Gender'].value_counts()

Male      489
Female    112
Name: Gender, dtype: int64

In [10]:
data['Gender'] = data['Gender'].replace({'male':'Male'})

In [11]:
print(f" There are {data['Gender'].value_counts()['Male']} male counts while there are {data['Gender'].value_counts()['Female']} female counts. Therefore, we will replace the missing values with male")

 There are 489 male counts while there are 112 female counts. Therefore, we will replace the missing values with male


In [12]:
data['Gender'] = data['Gender'].fillna('Male')

### Married

In [13]:
data['Married'].value_counts()

Yes    398
No     213
Name: Married, dtype: int64

In [14]:
data['Married'] = data['Married'].fillna('Yes')

### Dependents

In [15]:
data['Dependents'].value_counts()

0     345
1     102
2     101
3+     51
Name: Dependents, dtype: int64

In [16]:
data['Dependents'] = data['Dependents'].fillna(0)

### Self-employed

In [17]:
data['Self_Employed'].value_counts()

No     500
Yes     82
Name: Self_Employed, dtype: int64

In [18]:
data['Self_Employed'] = data['Self_Employed'].fillna('No')

### Loan Amount

we will fill the missing values with the average or mean of the sum of loan amouns.

In [19]:
data['LoanAmount'].value_counts()

120.0    20
110.0    17
100.0    15
160.0    12
187.0    12
         ..
240.0     1
214.0     1
59.0      1
166.0     1
253.0     1
Name: LoanAmount, Length: 203, dtype: int64

In [20]:
data['LoanAmount'] = data['LoanAmount'].fillna(data['LoanAmount'].mean())

### Loan Amount term

In [21]:
data['Loan_Amount_Term'].value_counts()

360.0    512
180.0     44
480.0     15
300.0     13
240.0      4
84.0       4
120.0      3
60.0       2
36.0       2
12.0       1
Name: Loan_Amount_Term, dtype: int64

In [22]:
data['Loan_Amount_Term'] = data['Loan_Amount_Term'].fillna(360)

### Credit History

In [23]:
data['Credit_History'].value_counts()

1.0    475
0.0     89
Name: Credit_History, dtype: int64

In [24]:
 data['Credit_History'].fillna(method ='ffill', inplace = True)

In [25]:
print(f"There is {data['Credit_History'].isnull().sum().sum()} missing value in the data.")

There is 0 missing value in the data.


In [26]:
data.head()

Gender Married Dependents     Education Self_Employed  ApplicantIncome  \
0   Male      No          0      Graduate            No             5849   
1   Male     Yes          1      Graduate            No             4583   
2   Male     Yes          0      Graduate           Yes             3000   
3   Male     Yes          0  Not Graduate            No             2583   
4   Male      No          0      Graduate            No             6000   

   CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
0                0.0  146.412162             360.0             1.0   
1             1508.0  128.000000             360.0             1.0   
2                0.0   66.000000             360.0             1.0   
3             2358.0  120.000000             360.0             1.0   
4                0.0  141.000000             360.0             1.0   

  Property_Area Loan_Status  
0         Urban           Y  
1         Rural           N  
2         Urban           Y  
3         Urban           Y  
4         Urban           Y

### Encoding Categorical values

In [27]:
df = data

In [28]:
data['Property_Area'].value_counts()

Semiurban    233
Urban        202
Rural        179
Name: Property_Area, dtype: int64

In [29]:
df['Gender'] = df['Gender'].map({'Male':0,'Female':1})
df['Married'] = df['Married'].map({'No':0,'Yes':1})
df['Education'] = df['Education'].map({'Not Graduate':0,'Graduate':1})
df['Self_Employed'] = df['Self_Employed'].map({'No':0,'Yes':1}) 

In [30]:
df['Property_Area'] = df['Property_Area'].map({'Urban':0,'Rural':1, 'Semiurban':2})
df['Dependents'] = df['Dependents'].map({'0':0, '1':1, '2':2, '3+':3})

In [31]:
df['Loan_Status'] = df['Loan_Status'].map({'N': 0, 'Y':1})

In [32]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 614 entries, 0 to 613
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Gender             614 non-null    int64  
 1   Married            614 non-null    int64  
 2   Dependents         599 non-null    float64
 3   Education          614 non-null    int64  
 4   Self_Employed      614 non-null    int64  
 5   ApplicantIncome    614 non-null    int64  
 6   CoapplicantIncome  614 non-null    float64
 7   LoanAmount         614 non-null    float64
 8   Loan_Amount_Term   614 non-null    float64
 9   Credit_History     614 non-null    float64
 10  Property_Area      614 non-null    int64  
 11  Loan_Status        614 non-null    int64  
dtypes: float64(5), int64(7)
memory usage: 57.7 KB


In [33]:
df = df.dropna()
df.head()

Gender  Married  Dependents  Education  Self_Employed  ApplicantIncome  \
0       0        0         0.0          1              0             5849   
1       0        1         1.0          1              0             4583   
2       0        1         0.0          1              1             3000   
3       0        1         0.0          0              0             2583   
4       0        0         0.0          1              0             6000   

   CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
0                0.0  146.412162             360.0             1.0   
1             1508.0  128.000000             360.0             1.0   
2                0.0   66.000000             360.0             1.0   
3             2358.0  120.000000             360.0             1.0   
4                0.0  141.000000             360.0             1.0   

   Property_Area  Loan_Status  
0              0            1  
1              1            0  
2              0            1  
3              0            1  
4              0            1

In [34]:
X = df.iloc[:,0:11].values
y = df.iloc[:,11].values.ravel()
y = np.array(y).astype(int)

In [35]:
# Splitting the dataset into the Training set and Test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

In [36]:
X_train

array([[  0.,   1.,   2., ..., 180.,   1.,   2.],
       [  0.,   1.,   0., ..., 360.,   1.,   1.],
       [  0.,   1.,   0., ..., 360.,   1.,   0.],
       ...,
       [  1.,   0.,   0., ..., 360.,   1.,   2.],
       [  0.,   1.,   1., ..., 360.,   1.,   2.],
       [  0.,   1.,   2., ..., 480.,   1.,   2.]])

In [37]:
# Feature Scaling
from sklearn import preprocessing
X_test = preprocessing.scale(X_test)
X_train = preprocessing.scale(X_train)
X = preprocessing.scale(X)

In [38]:
X_train

array([[-0.46114881,  0.72489841,  1.18117635, ..., -2.64778657,
         0.43426444,  1.09245268],
       [-0.46114881,  0.72489841, -0.76309675, ...,  0.2767863 ,
         0.43426444, -0.08877719],
       [-0.46114881,  0.72489841, -0.76309675, ...,  0.2767863 ,
         0.43426444, -1.27000706],
       ...,
       [ 2.1684974 , -1.37950364, -0.76309675, ...,  0.2767863 ,
         0.43426444,  1.09245268],
       [-0.46114881,  0.72489841,  0.2090398 , ...,  0.2767863 ,
         0.43426444,  1.09245268],
       [-0.46114881,  0.72489841,  1.18117635, ...,  2.22650154,
         0.43426444,  1.09245268]])

In [39]:
X_test

array([[-0.51298918, -1.31425748, -0.71044432, ...,  1.92262711,
        -2.23606798, -1.21574324],
       [ 1.94935887,  0.76088591, -0.71044432, ...,  0.29213062,
        -2.23606798, -1.21574324],
       [-0.51298918, -1.31425748, -0.71044432, ...,  0.29213062,
        -2.23606798,  0.04192218],
       ...,
       [-0.51298918,  0.76088591,  1.39457589, ..., -2.15361411,
        -2.23606798, -1.21574324],
       [-0.51298918, -1.31425748, -0.71044432, ...,  0.29213062,
         0.4472136 ,  1.29958761],
       [-0.51298918, -1.31425748,  0.34206578, ...,  0.29213062,
        -2.23606798,  1.29958761]])

# Deep Learning Neural Network

In [40]:
import tensorflow as tf
import keras
from keras.layers import Dense
from keras.optimizers import Adam

In [41]:
from keras.models import Sequential

### Constructing the network

In [42]:
y = y.reshape(-1, 1)

In [43]:
Seq_Net = Sequential()

In [44]:
Seq_Net.add(Dense(10, input_shape = (X.shape[1],), activation = 'relu', name = 'first'))
Seq_Net.add(Dense(20, activation ='relu', name = 'secong'))
Seq_Net.add(Dense(10, activation = 'tanh', name ='third'))
Seq_Net.add(Dense(6, activation ='sigmoid', name ='fourth'))
Seq_Net.add(Dense(y.shape[1], activation ='softmax', name ='output'))

In [45]:
optimizer = Adam(learning_rate = 0.001)
Seq_Net.compile(optimizer, loss ='categorical_crossentropy', metrics =['accuracy'])

In [46]:
print(Seq_Net.summary())

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 first (Dense)               (None, 10)                120       
                                                                 
 secong (Dense)              (None, 20)                220       
                                                                 
 third (Dense)               (None, 10)                210       
                                                                 
 fourth (Dense)              (None, 6)                 66        
                                                                 
 output (Dense)              (None, 1)                 7         
                                                                 
Total params: 623
Trainable params: 623
Non-trainable params: 0
_________________________________________________________________
None


### Fitting the model

In [47]:
Seq_Net.fit(X_train, y_train, verbose =1, batch_size = X.shape[1], epochs = 200)

Epoch 1/200
44/44 [==============================] - 9s 5ms/step - loss: 0.0000e+00 - accuracy: 0.6994
Epoch 2/200
44/44 [==============================] - 0s 5ms/step - loss: 0.0000e+00 - accuracy: 0.6994
Epoch 3/200
44/44 [==============================] - 0s 5ms/step - loss: 0.0000e+00 - accuracy: 0.6994
Epoch 4/200
44/44 [==============================] - 0s 5ms/step - loss: 0.0000e+00 - accuracy: 0.6994
Epoch 5/200
44/44 [==============================] - 0s 5ms/step - loss: 0.0000e+00 - accuracy: 0.6994
Epoch 6/200
44/44 [==============================] - 0s 5ms/step - loss: 0.0000e+00 - accuracy: 0.6994
Epoch 7/200
44/44 [==============================] - 0s 5ms/step - loss: 0.0000e+00 - accuracy: 0.6994
Epoch 8/200
44/44 [==============================] - 0s 5ms/step - loss: 0.0000e+00 - accuracy: 0.6994
Epoch 9/200
44/44 [==============================] - 0s 5ms/step - loss: 0.0000e+00 - accuracy: 0.6994
Epoch 10/200
44/44 [==============================] - 0s 5ms/step - loss:

### Testing the model

In [50]:
Seq_Net.evaluate(X_train, y_train)

15/15 [==============================] - 0s 4ms/step - loss: 0.0000e+00 - accuracy: 0.6994


[0.0, 0.699373722076416]

In [48]:
Seq_Net.evaluate(X_test, y_test)

4/4 [==============================] - 1s 5ms/step - loss: 0.0000e+00 - accuracy: 0.6500


[0.0, 0.6499999761581421]

In [49]:
print(Seq_Net.get_weights())

[array([[-9.9978194e-02,  4.9885529e-01, -1.1583936e-02, -7.3674873e-02,
        -9.5420532e-02,  5.0087982e-01, -4.3805301e-01, -2.9795064e-02,
        -1.9492668e-01, -3.8751218e-01],
       [ 5.9646212e-02, -3.6737227e-01, -2.6181358e-01, -2.4533805e-01,
         3.8587156e-01,  4.5794258e-01, -2.3359698e-01, -2.0621797e-01,
         2.0198368e-01,  1.6537802e-01],
       [-1.6940795e-01, -5.4147422e-01,  4.6287188e-01,  4.1040814e-01,
         5.2186501e-01,  4.2320055e-01, -3.3084550e-01, -4.3980348e-01,
         2.8975904e-01, -1.2077163e-01],
       [-4.0257055e-01, -5.4704279e-01,  2.4456728e-02, -4.2684221e-01,
         4.6730742e-01, -4.9486890e-02,  1.4874052e-02,  4.9332306e-01,
        -5.1690190e-04,  7.1938598e-01],
       [-6.5821427e-01,  4.7024471e-01, -4.8236811e-01, -3.7649494e-01,
         7.1277983e-02,  2.6377606e-01,  4.6847805e-02,  1.6165830e-01,
         3.5808668e-01, -4.2579690e-01],
       [-2.0962566e-01, -3.1325006e-01,  1.8289921e-01, -2.7371344e-01,
  

# Interpretation

The deep neural network model is used to predict loan. as you can see, 5 differents layers were used in this model. the first one represents the input layer, and the last one, the output layer. the second, third, and fourth represent the hidden layers. the evaluation of this model comes down to 65% accuracy in testing which is not too bad. in contrary to the training accuracy which is 69%. We can tell that the model is a perfect fit as both trained and test models generate accuracies level close to each other.